vnext/Microsoft.ReactNative/JsiApi.idl

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.ReactNative
+{
+  interface IJsiRuntime;
+
+  delegate void JsiDataHandler(UInt8[] data);
+
+  [webhosthidden]
+  interface IJsiBuffer
+  {
+    UInt32 Size { get; };
+    void GetData(JsiDataHandler handler);
+  };
+
+  enum JsiValueType
+  {
+    Undefined,
+    Null,
+    Boolean,
+    Number,
+    Symbol,
+    String,
+    Object,
+  };
+
+  [webhosthidden]
+  struct JsiValueData
+  {
+    JsiValueType Kind;
+    UInt64 Value;
+  };
+
+  [default_interface]
+  runtimeclass JsiPreparedJavaScript
+  {
+  };
+
+  delegate JsiValueData JsiHostFunction(IJsiRuntime runtime, JsiValueData thisValue, JsiValueData[] args);
+
+  [webhosthidden]
+  interface IJsiHostObject
+  {
+    JsiValueData GetProperty(IJsiRuntime runtime, UInt64 name);
+    void SetProperty(IJsiRuntime runtime, UInt64 name, JsiValueData value);
+    IVectorView<UInt64> GetPropertyNames(IJsiRuntime runtime);
+  };
+
+  [webhosthidden]
+  interface IJsiRuntime
+  {
+    JsiValueData EvaluateJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiPreparedJavaScript PrepareJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiValueData EvaluatePreparedJavaScript(JsiPreparedJavaScript js);
+    UInt64 Global { get; };
+    String Description { get; };
+    Boolean IsInspectable { get; };
+
+    UInt64 CloneSymbol(UInt64 symbol);
+    UInt64 CloneString(UInt64 str);
+    UInt64 CloneObject(UInt64 obj);
+    UInt64 ClonePropertyNameId(UInt64 propertyNameId);
+
+    UInt64 CreatePropertyNameIdFromAscii(UInt8[] ascii);
+    UInt64 CreatePropertyNameIdFromUtf8(UInt8[] utf8);
+    UInt64 CreatePropertyNameIdFromString(UInt64 str);
+    void PropertyNameIdToUtf8(UInt64 propertyNameId, JsiDataHandler utf8Handler);
+    Boolean ComparePropertyNameIds(UInt64 left, UInt64 right);
+
+    void SymbolToUtf8(UInt64 symbol, JsiDataHandler utf8Handler);
```

Comment:
Why not just return the data directly? Is there any benefit to use a callback?

Reply:
Yes, this is to avoid extra string copy and/or conversion. The callback based API allows to expose the underlying byte buffer directly without any copy.

---
In reply to: [451858417](https://github.com/microsoft/react-native-windows/pull/5461#discussion_r451858417) [](ancestors = 451858417)